### PR TITLE
SONARSLANG-125 SonarRuby: Basic AST mapping (Java version)

### PR DESCRIPTION
**Pull request by christophe-zurn-sonarsource**

### CODE REVIEW

File: sonar-ruby-plugin/src/test/java/org/sonarsource/ruby/converter/RubyConverterTest.java [hunk 1]
```diff
@@ -109,28 +91,26 @@ public void invalid_escape_sequence() {
     assertThat(converter.parse("\"\\xff\"")).isNotNull();
   }
 
-  @Test
-  public void top_level_tree() {
-    assertTree(converter.parse(("true\nfalse"))).isInstanceOf(TopLevelTree.class);
-    assertTree(converter.parse(("true\r\nfalse"))).isInstanceOf(TopLevelTree.class);
-  }
-
-  @Test
```
Comment (saberduck) [inline sonar-ruby-plugin/src/test/java/org/sonarsource/ruby/converter/RubyConverterTest.java]: Looks like this annotation was dropped accidentally, ranges is now not executed as test

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyConverter.java [hunk 2]
```diff
@@ -137,12 +139,28 @@ private Tree parseContent(String content) {
...
+    if (node == null || node.getTree() == null) {
+      // only comments
+      return new TopLevelTreeImpl(topTreeMetaData, emptyList(), comments);
+    } else {
+      // singleton expression: we wrap it around a top level tree
+      return new TopLevelTreeImpl(topTreeMetaData, singletonList(node.getTree()), comments);
+    }
+  }
+
+  private RubyVisitor getRubyProcessor(TreeMetaDataProvider metaDataProvider) {
```
Comment (saberduck) [inline sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyConverter.java]: Should we rename this method to `getRubyVisitor` ?

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+  @JRubyMethod(name = "new", required = 1, rest = true, meta = true)
+  public static IRubyObject rbNew(ThreadContext context, IRubyObject klazz, IRubyObject[] args) {
+    RubyVisitor rubyVisitor = (RubyVisitor) ((RubyClass) klazz).allocate();
+    rubyVisitor.metaDataProvider = args[0].toJava(TreeMetaDataProvider.class);
+    return rubyVisitor;
+  }
+
+  @JRubyMethod
+  public IRubyObject process(ThreadContext context, IRubyObject node) {
+    // some nodes present in the AST but are empty (e.g. function parameters clause)
```
Comment (saberduck) [inline sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java]: missing verb: some nodes **are** present ...

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    if (noLocation(node) || node.isNil()) {
+      return new RubyNil(getRuntime());
+    }
+
+    IRubyObject processedNode = callSuperMethod(context, "process", node);
+
+    if (processedNode instanceof NodeAdapter) {
+      // A more specific tree was already created by a specialized method (on_${type}, ex: on_int)
+      return processedNode;
+
```
Comment (saberduck) [inline sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java]: nitpick, remove whitespace

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    } else {
+      return createNativeTreeAdapter(processedNode);
+    }
+  }
+
+  @JRubyMethod(name = "on_class")
+  public IRubyObject onClass(ThreadContext context, IRubyObject node) {
+    IRubyObject updatedNode = callSuperMethod(context, "on_class", node);
```
Comment (saberduck) [inline sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java]: maybe we can rename `callSuperMethod`  to `continueVisit` , so it is more clear what it's doing

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    return convertToNodeAdapter(node, metaData -> new IdentifierTreeImpl(metaData, String.valueOf(name)));
+  }
+
+  @JRubyMethod(name = "on_int")
+  public IRubyObject onInt(ThreadContext context, IRubyObject node) {
+    Object value = getChild(node, 0);
+    return convertToNodeAdapter(node, metaData -> new LiteralTreeImpl(metaData, String.valueOf(value)));
+  }
+
+  private NodeAdapter convertToNodeAdapter(IRubyObject node, Function<TreeMetaData, Tree> getTree) {
```
Comment (saberduck) [inline sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java]: Can this be static factory method in `NodeAdapter` directly?

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+    RangeAdapter rangeAdapter = sourceMapAdapter.getRange();
+    return metaDataProvider.metaData(rangeAdapter.toTextRange());
+  }
+
+  private IRubyObject callSuperMethod(ThreadContext context, String methodName, IRubyObject arg1) {
+    DynamicMethod method = superclass.searchMethod(methodName);
+    return method.call(context, this, superclass, methodName, arg1);
+  }
+
+  private Object getChild(IRubyObject node, int index) {
```
Comment (saberduck) [inline sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java]: Can this method be moved to `NodeAdapter` as instance method?

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+    return ((List) JavaEmbedUtils.invokeMethod(getRuntime(), node, "to_a", null, List.class)).get(index);
+  }
+
+  private IRubyObject nodeType(IRubyObject node) {
+    return (IRubyObject) JavaEmbedUtils.invokeMethod(getRuntime(), node, "type", null, IRubyObject.class);
+  }
+
+  private List<Tree> getChildren(IRubyObject node, TreeMetaData metaData) {
```
Comment (saberduck) [inline sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java]: Can this method be moved to NodeAdapter as instance method?

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+  private IRubyObject callSuperMethod(ThreadContext context, String methodName, IRubyObject arg1) {
+    DynamicMethod method = superclass.searchMethod(methodName);
+    return method.call(context, this, superclass, methodName, arg1);
+  }
+
+  private Object getChild(IRubyObject node, int index) {
+    return ((List) JavaEmbedUtils.invokeMethod(getRuntime(), node, "to_a", null, List.class)).get(index);
+  }
+
+  private IRubyObject nodeType(IRubyObject node) {
```
Comment (saberduck) [inline sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java]: Can this method be moved to NodeAdapter as instance method?

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+      SourceMapAdapter sourceMapAdapter = new SourceMapAdapter(getRuntime(), location);
+      RangeAdapter rangeAdapter = sourceMapAdapter.getRange();
+
+      return rangeAdapter.isNull();
+    }
+
+    return false;
+  }
+
+  private TreeMetaData getMetaData(IRubyObject node) {
```
Comment (saberduck) [inline sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java]: Can this be moved to `NodeAdapter` ?

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/adapter/NodeAdapter.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+import org.jruby.RubyObject;
+import org.jruby.anno.JRubyClass;
+import org.jruby.anno.JRubyMethod;
+import org.jruby.javasupport.JavaEmbedUtils;
+import org.jruby.runtime.ThreadContext;
+import org.jruby.runtime.builtin.IRubyObject;
+import org.sonarsource.slang.api.Tree;
+
+@JRubyClass(name = "NodeAdapter")
+public class NodeAdapter extends RubyObject {
```
Comment (saberduck) [inline sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/adapter/NodeAdapter.java]: why it doesn't extend `JRubyObject` ?

File: slang-checks/src/main/java/org/sonarsource/slang/checks/BadFunctionNameCheck.java [hunk 12]
```diff
@@ -19,28 +19,42 @@
...
 
 @Rule(key = "S100")
 public class BadFunctionNameCheck implements SlangCheck {
 
-  private static final String DEFAULT_FORMAT = "^[a-z][a-zA-Z0-9]*$";
+  public static final String DEFAULT_FORMAT = "^[a-z][a-zA-Z0-9]*$";
+
+  private static final Map<String, String> DEFAULT_BY_LANGUAGE;
+  static {
+    HashMap<String, String> defaults = new HashMap<>();
```
Comment (saberduck) [inline slang-checks/src/main/java/org/sonarsource/slang/checks/BadFunctionNameCheck.java]: wdyt to have `enum` for languages in the `checks` module and use `EnumMap` here?

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/plugin/KotlinRulesDefinition.java [hunk 13]
```diff
@@ -35,10 +36,15 @@ public void define(Context context) {
...
       .setName(KotlinPlugin.REPOSITORY_NAME);
     RuleMetadataLoader ruleMetadataLoader = new RuleMetadataLoader(RESOURCE_FOLDER);
-// TODO: Add logic for rules that require language specific configuration at construction time
-    ArrayList<Class> checks = new ArrayList<>(CommonCheckList.kotlinChecks());
+
+    ArrayList<Class> checks = new ArrayList<>(CheckList.kotlinChecks());
     checks.add(CommentedCodeCheck.class);
     ruleMetadataLoader.addRulesByAnnotatedClass(repository, checks);
+
+    repository.rule("S100")
```
Comment (saberduck) [inline sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/plugin/KotlinRulesDefinition.java]: I think test is missing for default value

File: sonar-ruby-plugin/src/test/java/org/sonarsource/ruby/converter/RubyConverterTest.java (hunk 1 above)
Comment (christophe-zurn-sonarsource) [reply sonar-ruby-plugin/src/test/java/org/sonarsource/ruby/converter/RubyConverterTest.java]: Nice catch :+1:

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyConverter.java (hunk 2 above)
Comment (christophe-zurn-sonarsource) [reply sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyConverter.java]: Yep, to be consistent I renamed all remaining "rubyProcessor" to "rubyVisitor"

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java (hunk 3 above)
Comment (christophe-zurn-sonarsource) [reply sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java]: updated comment

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java (hunk 4 above)
Comment (christophe-zurn-sonarsource) [reply sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java]: done :)

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java (hunk 5 above)
Comment (christophe-zurn-sonarsource) [reply sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java]: Renamed method to `visitNode`

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java (hunk 6 above)
Comment (christophe-zurn-sonarsource) [reply sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java]: I'll leave this for now as is, and we'll improve that in another PR

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java (hunk 7 above)
Comment (christophe-zurn-sonarsource) [reply sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java]: I'll leave this for now as is, and we'll improve that in another PR

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java (hunk 8 above)
Comment (christophe-zurn-sonarsource) [reply sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java]: I'll leave this for now as is, and we'll improve that in another PR

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java (hunk 9 above)
Comment (christophe-zurn-sonarsource) [reply sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java]: I'll leave this for now as is, and we'll improve that in another PR

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java (hunk 10 above)
Comment (christophe-zurn-sonarsource) [reply sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/RubyVisitor.java]: I'll leave this for now as is, and we'll improve that in another PR

File: sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/adapter/NodeAdapter.java (hunk 11 above)
Comment (christophe-zurn-sonarsource) [reply sonar-ruby-plugin/src/main/java/org/sonarsource/ruby/converter/adapter/NodeAdapter.java]: Do you mean `IRubyObject`? `IRubyObject` is an interface, and we don't want to implement all base methods.
`RubyObject` is the equivalent of the ruby `Object` class, so classes that will be used inside the ruby runtime should extend this basic class.

File: slang-checks/src/main/java/org/sonarsource/slang/checks/BadFunctionNameCheck.java (hunk 12 above)
Comment (christophe-zurn-sonarsource) [reply slang-checks/src/main/java/org/sonarsource/slang/checks/BadFunctionNameCheck.java]: I'll leave this for now as is, and we'll improve that in another PR

File: sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/plugin/KotlinRulesDefinition.java (hunk 13 above)
Comment (christophe-zurn-sonarsource) [reply sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/plugin/KotlinRulesDefinition.java]: Added test for default value of this rule in Kotlin

**Comment by christophe-zurn-sonarsource** [comment]
@saberduck I've updated PR following your comments. I squashed the changes with their respective original commits so that we can keep the history on master when merging

**Pull request merged**